### ISSUE TITLE
How to activate animation when the content scrolls into view.

### ISSUE THREAD

**Issue by rakeshmali** [daneden/animate.css#508]
I have a some content and images that animates with animate.css and the animation currently activates as the page loads.

The problem I have is that the given content and images are placed off screen due to lots of content before it so by the time a user scrolls down to it, the animation has already finished.

I was looking for ways either through CSS3 or jQuery to only activate the animation on the content when the viewer sees the this.

If someone could point me in the right direction, it would be really helpful.

**Comment by Ooops-404** [comment]
Take a look at ScrollReveal
https://github.com/jlmakes/scrollReveal.js/tree/master.

**Comment by rakeshmali** [comment]
Is there any possibility for doing the same using animate.css? As I have already implemented it into my website :(

**Comment by FredyWenger** [comment]
First, thanks for your great work :-)
I'm starting with MVC6 and just implemented your .css in my MVC6 app, to see what happens. 
Principally it works and is easy to implement (how it should be).
But, I and also noted that problem (so... +1 to rakeshmali).
<strong>=> An animation only is useful, if the user is able to see it :-)</strong>

My feedback for now:
- principally very useful and easy to intergrate
- only usable in larger pages, if the animation is started, if the content comes to the viewable zone 
- works fine with small animated objects, that are viewable by start (in <strong>all</strong> browsers)
- I have some <strong>large</strong> pages where I show some large objects in <strong>bootstrap-</strong>objects:
  - only browser that shows the animation of the large object like expected: <strong>Firefox</strong>
  - in GoogleChrome, it stucks a bit
  - in IE11 and Edge, it don't run correct

I have done the tests with VS2015 in localhost (what is slow anyway), so I have to  test it on a selfhosting-host before I'm sure about the problems...


<strong>So my conclusion for now:</strong>
- <strong>Thanks once again for your work!</strong>
- If possible, your .css should handle, that the animation only is showed, if the content becomes 
  visible
  => Else it's too complicated to take care about by implement some further stuff

<strong>Only add a .css-file, that handles all itself is really "sexy"</strong> 

<strong>Can you say something about use the animations  with Bootstrap (is there somethin special to note about)?</strong>

**Comment by eltonmesquita** [comment]
@rakeshmali You'll have to use some javascript for this, there's no other way. Here a small jQuery function to do this: https://gist.github.com/eltonmesquita/bd803a21f27b12a58df1
If you prefer, I've done a better version with Vanilla Javascript + requestFrameAnimation for better performance, but it's more complex and have two small dependencies.

**Comment by eltonmesquita** [comment]
@FredyWenger I'm not sure if I understood everything, but here's some answers:
1- Animate.css is a CSS library, so there's a limit to what it can do. It can't, for instance, start an animation programatically as when an element is visible int the viewport, for this you have Javascript.
2 - You have to consider performance when using animations on an webpage/webapp. You shouldn't animate big elements as it can be costly. A good read is some of [Paul Lewis posts](https://aerotwist.com/blog/flip-your-animations/). It's worth to dig in his blog, you'll learn tons about performance.
3 - Animate.css should work fine with any framework, library or whatever you use, but be aware that CSS Animations use CSS property and it can override any property used in your selector, like if you use any transform in an element and add an animate.css class it'll probably override it and you'll have some problems. Another thing to be aware is that framework can have a lot of CSS for some elements and it can cause bad performance.

TL;DR: Animate.css takes care of the animations, how the animation is added and the good sense of using it performance wise, is up to you.

**Comment by FredyWenger** [comment]
@eltonmesquita:
Thanks for your work once again and your detailed answer.
1 - I have just started with MVC6 (and also CSS) but thought that there are some CSS "functions" that can be used to detect, if an element is viewable or not (and if yes, start the animation). I have created a new homepage for my company with wix.com. They have a stunning html-editor and various animations that can be configured easily (all run's only, it the element becomes visible and you can define, if they should run only once or every-time the element is showed). But they use reactjs and - maybe - do this by reactjs or javascript. 
As I have wrote, it would be sexy, if animate.css would take care of the problem itself (or you would "bundle" a matching javascript to animate.css so that the problem is solved within the bundled javascript and there where some properties link "animate one time" / "animate every time").  
But, O.K.... I have to look myself...
2 - Thanks for the link, I will check it out.
3 - So... it is no problem to use animale.css also with bootstrap.
TL,DR: That's clear and I will only use it where it make sense and do not downgrade the performance.

Finally thanks.
And... greetings from Switzerland.

**Comment by rakeshmali** [comment]
@eltonmesquita Thanks for sharing the trick of using animate.css according to my project. 
And @FredyWenger, you said it very true, "there should be properties link for uses as accordingly like "animate one time" / "animate every time"...
Animate.css do a great job, but have some disabilities in it's use cases. I think if they overcome from these disabilities then this will be the most popular/Useful library for all users [ Like me :) ].

**Comment by eltonmesquita** [comment]
To add some javascript to handle some common use cases may be a good idea but the problem is the this is out of the library's scope. You know, Animate.CSS.

If everything is clear, please close the issue.

**Comment by FredyWenger** [comment]
I have done some short tests (played a bit) with Javascript myself now (note: I work with MVC6 RC)
It seems to work with IE, Edge and Chrome, but not with FF.
So the animation is applied (one time), when the users scrolls.. 

<p>Test placeholder (Header)</p>
    <p>Test placeholder (Header)</p>
    <div class="animated fadeInLeft">
        <p><strong>Animation is done without check if visible or not</strong></p>
    </div>
    <p>Test Placeholder (Header)</p>
    <div class="test" id="test" onscroll='Animation();'>
        <p><strong>Animation with check if visible or not</strong></p>
    </div>
    @section scripts
    {
        <script type="text/javascript">
            {
                window.addEventListener("scroll", Animation);
                function Animation()
                {
                    $("#test:visible").addClass("animated fadeInLeft");
                }
            }
        </script>
    }

So, this set an event-handler to the window-scroll event and then apply the class, if the element becomes visible (and... don't work with FF).
Further, the case that the element is visible at startup is not handled yet and to take the scroll-event of the main-window may not be optimal...
But I'm sure, you could do it better... :-)

<strong>Anyway... thanks for your work...</strong>
<strong>For me, the case can be closed (but I'm not able to close it, as it isn't my case :-)</strong>



.

**Comment by FredyWenger** [comment]
@rakeshmali: Please close the case for  @eltonmesquita

**Comment by rakeshmali** [comment]
Ok thanks a lot Guys for your kind suggestions.

**Comment by jwhitleybb** [comment]
Is this JS defunct now? It's working for me. I'm using animate.css as well. http://jsfiddle.net/jarrodwhitley/h57ovjLb/114/

**Comment by nehadevli** [comment]
If you are using animate.css, then you have to include wow.js in your code.

Then animation will work as the text will be visible to viewport

**Comment by tomasts248** [comment]
What about using Aniview https://jjcosgrove.github.io/jquery-aniview/

**Comment by GordanShumway** [comment]
For all who still don't have a quick solution. I'm using it this way. The basic idea for scrolling came from Sam Sehnert, Digital Fusion:

`<!DOCTYPE html>
<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" href="animate.css">
<style>
body {
  margin: 100px
}

.slide-item-wrapper,
.slide-item-wrapper-up 
{
  width: 300px;
  height: 200px;
  overflow: hidden; 
}

.slide-item-wrapper-up 
{
margin-left: 250px;
}

.slide-item-wrapper > *,
.slide-item-wrapper-up > *
{
  transition: 2s ease-in-out;
}

.box-item {
  display: flex;
  justify-content: center;
  align-items: center;
  background: red;
  color: white;
  width: 300px;
  height: 200px;
  margin-bottom: 20px;
}

.animate
{
	opacity: 0;
}
</style>
</head>
<body>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js"></script>
<div class="slide-item-wrapper">
  <div>
    <div class="box-item animate --bounceInLeft">
      Lipsum
    </div>
     <div class="box-item">
      Lipsum2
    </div>
     <div class="box-item">
      Lipsum3
    </div>
     <div class="box-item">
      Lipsum4
    </div>
     <div class="box-item">
      Lipsum5
    </div>
     <div class="box-item">
      Lipsum6
    </div>
  </div>
</div>

<br>
<br>
<br>

<div class="slide-item-wrapper-up animate --bounceInRight">
  <div>
    <div class="box-item ">
      Lipsum
    </div>
     <div class="box-item">
      Lipsum6
    </div>
     <div class="box-item">
      Lipsum5
    </div>
     <div class="box-item">
      Lipsum4
    </div>
     <div class="box-item">
      Lipsum3
    </div>
     <div class="box-item">
      Lipsum2
    </div>
  </div>
</div>

<br>
<br>
<br>

<div class="slide-item-wrapper2">
  <div>
    <div class="box-item">
      Lipsum
    </div>
     <div class="box-item">
      Lipsum2
    </div>
     <div class="box-item animate --bounceInRight">
      Lipsum3
    </div>
     <div class="box-item">
      Lipsum4
    </div>
     <div class="box-item">
      Lipsum5
    </div>
     <div class="box-item animate --bounce">
      Lipsum6
    </div>
  </div>
</div>

<script>

	var win = jQuery(window);
	var slideUpMods = jQuery('.animate');
	
	function IsElementVisible (element, partial)
	{
		var jQueryt           = jQuery(element),
			jQueryw       = jQuery(window),
			viewTop       = jQueryw.scrollTop(),
			viewBottom    = viewTop + jQueryw.height(),
			_top          = jQueryt.offset().top,
			_bottom       = _top + jQueryt.height(),
			compareTop    = partial === true ? _bottom : _top,
			compareBottom = partial === true ? _top : _bottom;
    
		return ((compareBottom <= viewBottom) && (compareTop >= viewTop));
	}
	
	function runAnimations()
	{
		var isMobile = window.innerWidth < 768;		
		
		slideUpMods = jQuery('.animate');
		
  		slideUpMods.each(function(i, el) 
		{
			var el = jQuery(el);
			
			if (isMobile)
			{
				el.css('opacity', '1.0');
			}
			else
			{						
				if (IsElementVisible(el, false))
				{
					var classNames = el.attr("class").split(' ');
					
					for (var i = 0; i <= classNames.length - 1; i++)
					{
						var clName = classNames[i];
						if (clName.startsWith('--'))
						{
							el.removeClass(clName);
							el.removeClass('animate');
							clName = clName.replace('--', '');
							el.addClass('animated'); 
							el.addClass(clName); 
						}
					};
				} 
			}
		})
	}
	
	runAnimations();

    win.scroll(function(event) 
	{  
		runAnimations();
  	});
</script>

</body>
</html> `

**Comment by jarrodwhitley** [comment]
wow.js works perfectly for me.

**Comment by GordanShumway** [comment]
Wow! Thank you! Very good tip!

**Comment by 92Eli** [comment]
> @rakeshmali You'll have to use some javascript for this, there's no other way. Here a small jQuery function to do this: https://gist.github.com/eltonmesquita/bd803a21f27b12a58df1

Isn't it bad practice to have javascript constantly running every time the user scrolls?

**Comment by amitprajapati2298** [comment]
> For all who still don't have a quick solution. I'm using it this way. The basic idea for scrolling came from Sam Sehnert, Digital Fusion. This sample is quick and dirty, but ready to copy&paste and try...
> 
> ```
> <!DOCTYPE html>
> <html>
> <head>
> <meta name="viewport" content="width=device-width, initial-scale=1">
> <link rel="stylesheet" href="animate.css">
> <style>
> body {
>   margin: 100px
> }
> 
> .slide-item-wrapper,
> .slide-item-wrapper-up 
> {
>   width: 300px;
>   height: 200px;
>   overflow: hidden; 
> }
> 
> .slide-item-wrapper-up 
> {
> margin-left: 250px;
> }
> 
> .slide-item-wrapper > *,
> .slide-item-wrapper-up > *
> {
>   transition: 2s ease-in-out;
> }
> 
> .box-item {
>   display: flex;
>   justify-content: center;
>   align-items: center;
>   background: red;
>   color: white;
>   width: 300px;
>   height: 200px;
>   margin-bottom: 20px;
> }
> 
> .animate
> {
> 	opacity: 0;
> }
> </style>
> </head>
> <body>
> <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.1/jquery.min.js"></script>
> <div class="slide-item-wrapper">
>   <div>
>     <div class="box-item animate --bounceInLeft">
>       Lipsum
>     </div>
>      <div class="box-item">
>       Lipsum2
>     </div>
>      <div class="box-item">
>       Lipsum3
>     </div>
>      <div class="box-item">
>       Lipsum4
>     </div>
>      <div class="box-item">
>       Lipsum5
>     </div>
>      <div class="box-item">
>       Lipsum6
>     </div>
>   </div>
> </div>
> 
> <br>
> <br>
> <br>
> 
> <div class="slide-item-wrapper-up animate --bounceInRight">
>   <div>
>     <div class="box-item ">
>       Lipsum
>     </div>
>      <div class="box-item">
>       Lipsum6
>     </div>
>      <div class="box-item">
>       Lipsum5
>     </div>
>      <div class="box-item">
>       Lipsum4
>     </div>
>      <div class="box-item">
>       Lipsum3
>     </div>
>      <div class="box-item">
>       Lipsum2
>     </div>
>   </div>
> </div>
> 
> <br>
> <br>
> <br>
> 
> <div class="slide-item-wrapper2">
>   <div>
>     <div class="box-item">
>       Lipsum
>     </div>
>      <div class="box-item">
>       Lipsum2
>     </div>
>      <div class="box-item animate --bounceInRight">
>       Lipsum3
>     </div>
>      <div class="box-item">
>       Lipsum4
>     </div>
>      <div class="box-item">
>       Lipsum5
>     </div>
>      <div class="box-item animate --bounce">
>       Lipsum6
>     </div>
>   </div>
> </div>
> 
> <script>
> 
> 	var win = jQuery(window);
> 	var slideUpMods = jQuery('.animate');
> 	
> 	function IsElementVisible (element, partial)
> 	{
> 		var jQueryt           = jQuery(element),
> 			jQueryw       = jQuery(window),
> 			viewTop       = jQueryw.scrollTop(),
> 			viewBottom    = viewTop + jQueryw.height(),
> 			_top          = jQueryt.offset().top,
> 			_bottom       = _top + jQueryt.height(),
> 			compareTop    = partial === true ? _bottom : _top,
> 			compareBottom = partial === true ? _top : _bottom;
>     
> 		return ((compareBottom <= viewBottom) && (compareTop >= viewTop));
> 	}
> 	
> 	function runAnimations()
> 	{
> 		var isMobile = window.innerWidth < 768;		
> 		
> 		slideUpMods = jQuery('.animate');
> 		
>   		slideUpMods.each(function(i, el) 
> 		{
> 			var el = jQuery(el);
> 			
> 			if (isMobile)
> 			{
> 				el.css('opacity', '1.0');
> 			}
> 			else
> 			{						
> 				if (IsElementVisible(el, false))
> 				{
> 					var classNames = el.attr("class").split(' ');
> 					
> 					for (var i = 0; i < classNames.length ; i++)
> 					{
> 						var clName = classNames[i];
> 						if (clName.startsWith('--'))
> 						{
> 							el.removeClass(clName);
> 							el.removeClass('animate');
> 							clName = clName.replace('--', '');
> 							el.addClass('animated'); 
> 							el.addClass(clName); 
> 						}
> 					};
> 				} 
> 			}
> 		})
> 	}
> 	
> 	runAnimations();
> 
>     win.scroll(function(event) 
> 	{  
> 		runAnimations();
>   	});
> </script>
> 
> </body>
> </html> 
> ```

I have used Code. It is working very well. Thank you for providing this code
